src/h2_stream_set.erl

```diff
@@ -261,51 +349,160 @@ new_stream(
+            Settings
+    catch error:badarg ->
+              #settings{}
+    end.
+
+get_self_settings_locked(StreamSet=#stream_set{table=T}) ->
+    SelfSettings = get_self_settings(StreamSet),
+    try ets:select_replace(T, [{#connection_settings{type=self_settings, settings=SelfSettings}, [], [{const, #connection_settings{type=self_settings, settings=locked}}]}]) of
+        1 ->
+            SelfSettings;
+        0 ->
+            timer:sleep(1),
+            get_self_settings_locked(StreamSet)
+    catch
+        error:badarg ->
+            #settings{}
+    end.
+
+get_peer_settings_locked(StreamSet=#stream_set{table=T}) ->
+    PeerSettings = get_peer_settings(StreamSet),
+    try ets:select_replace(T, [{#connection_settings{type=peer_settings, settings=PeerSettings}, [], [{const, #connection_settings{type=peer_settings, settings=locked}}]}]) of
+        1 ->
+            PeerSettings;
+        0 ->
+            timer:sleep(1),
+            get_peer_settings_locked(StreamSet)
+    catch
+        error:badarg ->
+            #settings{}
+    end.
+
+get_settings(StreamSet) ->
+    {get_self_settings(StreamSet), get_peer_settings(StreamSet)}.
+
+update_self_settings(#stream_set{table=T}, Settings) ->
+    try ets:select_replace(T, [{#connection_settings{type=self_settings, settings=locked}, [], [{const, #connection_settings{type=self_settings, settings=Settings}}]}]) of
+        1 -> ok
+    catch 
+        error:badarg ->
+              ok
+    end.
+
+update_peer_settings(#stream_set{table=T}, Settings) ->
+    try ets:select_replace(T, [{#connection_settings{type=peer_settings, settings=locked}, [], [{const, #connection_settings{type=peer_settings, settings=Settings}}]}]) of
+        1 -> ok
+    catch
+        error:badarg ->
+            ok
+    end.
+
+get_encode_context(StreamSet) ->
+    get_encode_context(StreamSet, force).
+
+get_encode_context(StreamSet, Headers) ->
+    try (hd(ets:lookup(StreamSet#stream_set.table, encode_context))) of
+        #context{context=locked} ->
+            timer:sleep(1),
+            get_encode_context(StreamSet, Headers);
+        Encoder=#context{context=EncodeContext} ->
+            case Headers /= force andalso hpack:all_fields_indexed(Headers, EncodeContext) of
+                true ->
+                    %% we don't have any new headers, so we don't need the lock
+                    {nolock, EncodeContext};
+                false ->
+                    case ets:select_replace(StreamSet#stream_set.table, [{Encoder, [], [{const, #context{type=encode_context, context=locked}}]}]) of
+                        1 ->
+                            %% ok we have the ownership of this now
+                            {lock, EncodeContext};
+                        0 ->
+                            timer:sleep(1),
+                            get_encode_context(StreamSet, Headers)
+                    end
+            end
+    catch
+        error:badarg ->
+            hpack:new_context()
```

Comment:
Everywhere calling this is expecting a 2-tuple. `{Lock, EncodeContext}`

Reply:
Ah yup, needs to return that 2-tuple. Guessing `{nolock, hpack:new_context()}` @Vagabond ?

Reply:
yep, although really this is just a shim to quiet errors when the connection is closing